### PR TITLE
datetime: fix strptime for fuzzer

### LINKED ISSUE #8525: datetime_strptime_fuzzer: ASSERT: mday >= 1 && mday <= 31
**Bug description**

Fuzzing test for `datetime_strptime` is broken by asserts:

https://github.com/tarantool/tarantool/blob/abb172a245a80e51a81b545f8d5d34cc21023271/src/lib/core/datetime.c#L148-L149

```
INFO: Running with entropic power schedule (0xFF, 100).
  | INFO: Seed: 3480450009
  | INFO: Loaded 1 modules   (34816 inline 8-bit counters): 34816 [0x93e648, 0x946e48),
  | INFO: Loaded 1 PC tables (34816 PCs): 34816 [0x946e48,0x9cee48),
  | /mnt/scratch0/clusterfuzz/bot/builds/clusterfuzz-builds_tarantool_c6a08d9c704777ebe38728ec73f5e6d849eea0cd/revisions/datetime_strptime_fuzzer: Running 1 inputs 100 time(s) each.
  | Running: /mnt/scratch0/clusterfuzz/bot/inputs/fuzzer-testcases/8ff941e0d3ea7116167dc4eb58f808a1083a1870e6ebd72d608117d18071fefe
  | datetime_strptime_fuzzer: ./src/lib/core/datetime.c:148: _Bool tm_to_datetime(struct tnt_tm *, struct datetime *): Assertion `mday >= 1 && mday <= 31' failed.
```

We need to replace these assert's to `if`..`then`.

Reproducer: 
[clusterfuzz-testcase-minimized-datetime_strptime_fuzzer-6575551489114112.txt](https://github.com/tarantool/tarantool/files/11136545/clusterfuzz-testcase-minimized-datetime_strptime_fuzzer-6575551489114112.txt)

https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=57602

### PR DESCRIPTION
**Pull request by tsafin**
Google fuzzing efforts revealed yet another bound condition we don't handle well in the `tnt_strptime` function:

- for format `%m%g%W`;
- and input string `07001`.

We failed with assertion failure:
```
| datetime_strptime_fuzzer: ./src/lib/core/datetime.c:148: \
_Bool tm_to_datetime(struct tnt_tm *, struct datetime *): \
Assertion `mday >= 1 && mday <= 31' failed.
```

Closes #8525

### CODE REVIEW

**Review by ligurio — COMMENTED** [review]
I've checked that patch fixes an assertion caught by fuzzing test. From a first glance, the patch is correct, however I didn't dig to a code of `tnt_strptime` enough to make sure that fix doesn't bring more problems than it fixes.

Also added a missed changelog entry for fix in a separate commit.

File: changelogs/unreleased/gh-8525-strptime-fix.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## bugfix/datetime
+
+* Fixed a bug in `strptime` when assertion was triggered (gh-8525).
```
Comment (andreyaksenov) [inline changelogs/unreleased/gh-8525-strptime-fix.md:3]:
```suggestion
* Fixed a bug in `strptime` when the assertion was triggered (gh-8525).
```

**Review by andreyaksenov — APPROVED** [review]

File: changelogs/unreleased/gh-8525-strptime-fix.md (hunk 1 above)
Comment (ligurio) [reply changelogs/unreleased/gh-8525-strptime-fix.md]: Fixed, thanks!

**Review by ligurio — APPROVED** [review]
LGTM

**Review by ochaplashkin — APPROVED** [review]

**Comment by igormunkin** [comment]
Also backported to 3.0 and 2.11 in scope of e73ef15 and cd32470 respectively.

**Pull request merged**